### PR TITLE
feat: store API keys in the system keyring

### LINKED ISSUE #130: Store API keys in the system keyring instead of plaintext
API keys are currently stored in plaintext at `~/.config/linear/credentials.toml`. Anyone with filesystem access can read them. This is the same problem `gh` had before adding `--secure-storage`. It is a common vulnerability in CLIs, but nonetheless worth addressing. Especially when the credentials are not time-bound, as is the case with Linear API keys. 

## Proposed Approach

Shell out to platform keyring CLIs:

- macOS: `/usr/bin/security` (`add-generic-password` / `find-generic-password`)
- Linux: `secret-tool` (libsecret/D-Bus)
- Windows: PowerShell `CredentialManager`

This is the same approach `gh` uses via [zalando/go-keyring](https://github.com/zalando/go-keyring). It avoids native dependencies and does not require code signing or Apple notarization, since the Deno binary delegates to Apple's own `security` binary rather than calling Security framework APIs directly.

The tradeoff vs. direct API access (see [terraform-credentials-keychain](https://github.com/bendrucker/terraform-credentials-keychain) as an example project that I did the hard way) is that the macOS Keychain trust dialog shows "security" as the requesting app rather than "linear". Seems acceptable for this project given the setup costs of notarization (including Apple Developer membership).

## Credential Storage Layout

`credentials.toml` currently stores both metadata (workspace list, default) and secrets (API keys) in one file:

```toml
default = "my-workspace"
my-workspace = "lin_api_xxxxx"
```

With keyring, split the concerns:
- **Keychain**: stores API keys, keyed as `linear-cli:<workspace-slug>`
- **`credentials.toml`**: stores only metadata — which workspaces exist and which is default

```toml
default = "my-workspace"
workspaces = ["my-workspace", "other-workspace"]
```

## Code Changes

The credential layer is already well-isolated in `src/credentials.ts`. All reads/writes go through functions like `addCredential()`, `getCredentialApiKey()`, etc. The changes:

- New `src/keyring.ts` module: platform-detecting wrapper with `getPassword(service, account)`, `setPassword(...)`, `deletePassword(...)`, shelling out via `Deno.Command`
- Update `src/credentials.ts`: read/write API keys through keyring instead of TOML values, keep TOML for workspace metadata only
- Update `auth login` / `auth logout` — store/delete keyring entries
- Update `auth list` — validate keyring entries are accessible

## Headless / CI Environments

When no keyring is available (SSH, Docker, CI), `LINEAR_API_KEY` env var is already the recommended path and takes highest priority in the token resolution chain. No fallback to file-based secret storage needed, if the keyring is unavailable and no env var is set, error with a message pointing to `LINEAR_API_KEY`.

## Out of Scope

Ideas I considered but would skipping include:

- **Migration Tooling**: No auto-migration of existing plaintext credentials. Users re-run `auth login` after upgrading. Can be documented in release notes.
- **Native Keyring Libraries** (`@napi-rs/keyring`, Deno FFI): Unnecessary complexity. The shell-out approach is proven by `gh` and avoids the signing/notarization pipeline.
- **Opt-In**: Make keyring the default, not opt-in. The env var path already covers environments without a keyring.

## Open Questions

- On Linux, `secret-tool` may not be installed. Should the CLI suggest installing it, or fall back to file storage on Linux specifically?

### PR DESCRIPTION
**Pull request by bendrucker**
Move API key storage from plaintext TOML to OS-native keyrings (macOS Keychain, Linux `libsecret`, Windows Credential Manager). The credentials file retains only workspace metadata. Keys are loaded into an in-memory cache at startup so all downstream reads remain synchronous — no changes needed to any command files.

### Keyring (`src/keyring/`)

- Platform-detecting wrapper with `getPassword`, `setPassword`, `deletePassword` exports
- macOS: `/usr/bin/security` (exit 44 = not found)
- Linux: `secret-tool` via stdin for writes (exit 1 = not found)
- Windows: `Deno.dlopen("advapi32.dll")` FFI calling `CredReadW`/`CredWriteW`/`CredDeleteW` directly
- `_setBackend()` test seam for injecting an in-memory `Map` backend

### Windows Credential Manager via FFI

The Windows backend calls `advapi32.dll` directly via Deno's FFI (`Deno.dlopen`) rather than shelling out to PowerShell. This matches the standard approach taken by every comparable credential tool:

- [`danieljoos/wincred`](https://github.com/danieljoos/wincred/blob/623325312d3224d48d131159187b93e906216563/sys.go) — Go library calling `advapi32.dll` via `windows.NewLazySystemDLL`, used by:
  - [`docker-credential-helpers`](https://github.com/docker/docker-credential-helpers/blob/2b4e08bca3dbdb8e6c6e28790042742d0c0fc48f/wincred/wincred.go)
  - [`gh` CLI](https://github.com/cli/cli/blob/2c54a0d36a2f3c9c1f1b869a64120837c3a1e6f5/internal/keyring/keyring.go) (via [`zalando/go-keyring`](https://github.com/zalando/go-keyring/blob/5c6f7e0ba54d20daa8ea4e03f7ce0a27c075bfb6/keyring_windows.go))
  - [`aws-vault`](https://github.com/99designs/aws-vault/blob/70522e8f0b8f9c5b4e2e4e1e1e1cc4e3e5c3f04c/go.mod) (via `99designs/keyring`)
- [`node-keytar`](https://github.com/atom/node-keytar/blob/deae59a488789f2cd4a8dba6c7e58665795804fe/src/keytar_win.cc) — C++ N-API addon, `#include <wincred.h>`
- [`jaraco/keyring`](https://github.com/jaraco/keyring/blob/38c040133559682902f25fe96496756ee6849820/keyring/backends/Windows.py) — Python, `win32cred` (pywin32-ctypes wrapping advapi32 via ctypes)

The implementation packs the 80-byte `CREDENTIALW` struct manually via `DataView`, encodes strings as UTF-16LE for the `W`-suffix APIs, and uses `GetLastError` from `kernel32.dll` to distinguish "not found" (`ERROR_NOT_FOUND` = 1168) from real failures. DLLs are lazy-loaded so the module import doesn't fail on macOS/Linux.

### Credentials (`src/credentials.ts`)

- `Credentials` interface changed from index signature to `{ default?: string; workspaces: string[] }`
- `apiKeyCache` `Map` populated at startup, keeping `getCredentialApiKey()` sync
- `addCredential`/`removeCredential` write to keyring first, only mutate local state on success
- `parseInlineCredentials` / `parseKeyringCredentials` / `populateKeyringCache` extracted from `loadCredentials`
- Parallel keyring lookups via `Promise.all`
- Malformed TOML parse errors caught with recovery guidance
- Warnings for: missing keyring entries, dangling default workspace, inline format detected

### Backward Compatibility

- Inline-format TOML files (keys stored as `workspace = "lin_api_..."`) are detected by `hasInlineKeys` and served from the file directly
- `addCredential` on an inline-format installation rewrites the file to keyring format

### Auth List (`src/commands/auth/auth-list.ts`)

- Replaces removed `getAllCredentials()` with `getApiKeyForWorkspace()`
- Distinguishes auth errors (401/403) from network/other failures instead of labeling everything "invalid credentials"

### CI

- Added `keyring-integration` job on `macos-latest` and `windows-latest` for real credential round-trip testing

## Testing

- Subprocess isolation via `deno eval` for credential tests (required by top-level `await loadCredentials()`)
- Mock keyring backend injected via `_setBackend` — covers happy paths, error propagation, and cache consistency
- Integration test (`test/keyring.integration.test.ts`) exercises the real macOS Keychain and Windows Credential Manager lifecycle
- Edge cases covered: keyring write/delete failures leave state unchanged, null keyring returns warn but don't crash, dangling default dropped on load, inline→keyring format transition on `addCredential`

## References

Closes #130

### CODE REVIEW

File: src/keyring.ts [hunk 1]
```diff
@@ -0,0 +1,274 @@
...
+      throw new Error(`Unsupported platform: ${Deno.build.os}`)
+  }
+}
+
+function platformHint(): string {
+  switch (Deno.build.os) {
+    case "darwin":
+      return "Could not find /usr/bin/security. Is this a macOS system?"
+    case "linux":
+      return "Could not find secret-tool. Install it with: apt install libsecret-tools\n" +
```
Comment (schpet) [inline src/keyring.ts]: this assumes debian/ubuntu - i don't want to disappoint arch users etc

File: src/keyring.ts [hunk 2]
```diff
@@ -0,0 +1,274 @@
...
+      throw new Error(
+        `secret-tool clear failed (exit ${result.code}): ${result.stderr}`,
+      )
+    }
+  },
+}
+
+function escapePowerShell(s: string): string {
+  return s.replace(/'/g, "''")
+}
```
Comment (schpet) [inline src/keyring.ts]: is there a deno std thing that might help with this? this looks fragile

**Comment by schpet** [comment]
@bendrucker fyi i finally was able to release existing stuff, so i'm going to QA this next. hoping to find time this week, having the bots do some prelim investigation now. but this one might take me a bit to review (i would ideally like to make sure it works on windows but don't have a windows machine, etc etc)

File: src/credentials.ts
Comment (schpet) [inline src/credentials.ts]: i'm logged into a few workspaces, i tried running a command and got this:

```
Warning: Credentials file uses inline plaintext format. Run `linear auth login` for each workspace to migrate to the system keyring.
```

so then i logged into one of my workspaces, but it deleted all my other api keys so i was logged out of all of them. ux wise we could improve that, some sort of auto-migrate or whatever, but overall i think i'd rather keep supporting the toml credentials, too.

what do you think about this:

a) continue to accept credentials-on-disk if they are stored there, without printing a warning, i.e. completely backwards compatible
b) when setting a credential, support both the netrc style 'write credentials to a toml file lol' in addition to the keyring/secret manager - feature detect the secret manager and default to it if it's there, otherwise fallback to toml, but let cli flags override this

**Review by schpet — CHANGES_REQUESTED** [review]

**Comment by bendrucker** [comment]
Updated to address discussion from the last review. The main changes since 85eea5f:

Backward-compatible credential storage (3cf4b7c, fa97796):

- Inline (plaintext) TOML credentials are now fully supported without warnings
- `addCredential` preserves whatever format is currently in use. If your file has inline keys, new logins stay inline
- `removeCredential` and `setDefaultWorkspace` also preserve the format (these had the same silent-conversion bug)
- Passing `--plaintext` explicitly stores in TOML; passing `--no-plaintext` migrates all keys to keyring
- `saveInlineCredentials` throws instead of silently dropping workspaces with missing cache entries

Keyring opt-in with feature detection (0ca140c):

- `linear auth login` checks keyring availability before attempting keyring storage. If no keyring and no `--plaintext` flag, it errors with guidance
- When logging in with existing inline credentials, a note is printed and the user is prompted to migrate to keyring (if available)
- `linear auth status` now shows the credential storage type (plaintext file vs system keyring) and keyring availability
- `linear auth migrate` added as an explicit non-interactive command to migrate plaintext credentials to keyring

Migration safety (fa97796):

- `migrateToKeyring` rolls back already-written keyring entries on partial failure (best effort) and leaves the inline file untouched
- `keyring.isAvailable()` now logs the error detail instead of silently swallowing it

Other cleanup:

- Inner catch in `auth login` re-throws `CliError`/`AuthError`/`ValidationError` instead of wrapping them in a generic error
- New tests covering format preservation, migration, rollback, and the migrate command

**Comment by schpet** [comment]
@bendrucker i tested it on mac: worked without a hitch. i learned the linux secret managers are generally gnome so wasn't able to try it out on the headless vms i generally work on, so hoping that works or someone lets me know if it doesn't 🤞🏻 

so thanks again for this, what a great improvement.

**Comment by bendrucker** [comment]
Have found the same re: testing on Linux with keyrings, happy to help if someone is able to try it properly there and finds any issues
